src/auth/ntlm.js

```diff
@@ -0,0 +1,276 @@
+const crypto = require('crypto');
+const BigInteger = require('big-number').n;
+
+const hex = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'];
+
+const NTLMFlags = {
+  NTLM_NegotiateUnicode: 0x00000001,
+  NTLM_NegotiateOEM: 0x00000002,
+  NTLM_RequestTarget: 0x00000004,
+  NTLM_Unknown9: 0x00000008,
+  NTLM_NegotiateSign: 0x00000010,
+  NTLM_NegotiateSeal: 0x00000020,
+  NTLM_NegotiateDatagram: 0x00000040,
+  NTLM_NegotiateLanManagerKey: 0x00000080,
+  NTLM_Unknown8: 0x00000100,
+  NTLM_NegotiateNTLM: 0x00000200,
+  NTLM_NegotiateNTOnly: 0x00000400,
+  NTLM_Anonymous: 0x00000800,
+  NTLM_NegotiateOemDomainSupplied: 0x00001000,
+  NTLM_NegotiateOemWorkstationSupplied: 0x00002000,
+  NTLM_Unknown6: 0x00004000,
+  NTLM_NegotiateAlwaysSign: 0x00008000,
+  NTLM_TargetTypeDomain: 0x00010000,
+  NTLM_TargetTypeServer: 0x00020000,
+  NTLM_TargetTypeShare: 0x00040000,
+  NTLM_NegotiateExtendedSecurity: 0x00080000,
+  NTLM_NegotiateIdentify: 0x00100000,
+  NTLM_Unknown5: 0x00200000,
+  NTLM_RequestNonNTSessionKey: 0x00400000,
+  NTLM_NegotiateTargetInfo: 0x00800000,
+  NTLM_Unknown4: 0x01000000,
+  NTLM_NegotiateVersion: 0x02000000,
+  NTLM_Unknown3: 0x04000000,
+  NTLM_Unknown2: 0x08000000,
+  NTLM_Unknown1: 0x10000000,
+  NTLM_Negotiate128: 0x20000000,
+  NTLM_NegotiateKeyExchange: 0x40000000,
+  NTLM_Negotiate56: 0x80000000
+};
+
+const DEFAULT_NEGOTIATE_FLAGS = (
+  NTLMFlags.NTLM_NegotiateUnicode |
+  NTLMFlags.NTLM_NegotiateOEM |
+  NTLMFlags.NTLM_RequestTarget |
+  NTLMFlags.NTLM_NegotiateNTLM |
+  NTLMFlags.NTLM_NegotiateAlwaysSign |
+  NTLMFlags.NTLM_NegotiateExtendedSecurity |
+  NTLMFlags.NTLM_Negotiate128 |
+  NTLMFlags.NTLM_Negotiate56
+) >>> 0;
+
+/**
+  Authenticate to SQL Server via NTLM.
+
+  This class allows authentication to SQL Server via NTLM, using
+  target system or domain credentials.
+*/
+class NTLMAuthProvider {
+  constructor(connection, options) {
+    this.connection = connection;
+    this.options = options;
+  }
+
+  handshake(data, callback) {
+    if (!data) {
+      process.nextTick(callback, null, this.buildNegotiateMessage());
+    } else {
+      process.nextTick(callback, null, this.buildAuthenticateMessage(this.parseChallengeMessage(data)));
+    }
+  }
+
+  buildNegotiateMessage() {
+    const messageBuffer = Buffer.alloc(40);
+    const domainBuffer = Buffer.from(this.options.domain || '', 'ascii');
+    const workstationBuffer = Buffer.from(this.options.workstation || '', 'ascii');
+
```

Comment:
`config.workstation` is not re-assigned to `this.config.workstation` in connection.js (referring to master branch), was it missed somehow in the early PRs? or was that intentional? Can you give me some context?

Reply:
The `workstation`  option was never exposed via the options `Connection`, so there is no need to expose it now.

If a user wants to specify this option, they will be able to install the `tedious-auth-ntlm` module and pass that option manually, like this:

```js
const ntlmAuthProvider = require('tedious-auth-ntlm');

const connection = new Connection({
  ...
  authProvider: ntlmAuthProvider({
    domain: 'DOMAIN',
    workstation: 'WORKSTATION',
    username: 'username',
    password: 'password'
  }),
  ...
});
```